### PR TITLE
`LocalChain` with hardwired genesis block

### LINKED ISSUE #1079: Create LocalChain with genesis block hardwired for Network
**Describe the enhancement**

`LocalChain` will no longer implement `Default`. Instead, it will require a genesis hash to construct. This ensures that we will always have at least one checkpoint in `LocalChain` (the genesis block).

`ChainOracle::get_chain_tip` will no longer need to return an `Option` as we guarantee that we at least have the genesis block.

```rust
pub trait ChainOracle {
    /* other trait methods and types */

    /// Get the best chain's chain tip.
    fn get_chain_tip(&self) -> Result<BlockId, Self::Error>;
}
```

**Implementation details**

`LocalChain` will need to ensure that the genesis checkpoint cannot be replaced while updating. I think it would also make sense to persist the genesis checkpoint, this way recovery from persistence will fail with the wrong genesis hash.

**Changes to chain-source crates**

Any method that takes in an option of a checkpoint should just take in the checkpoint (no `Option`).

For the `bdk_bitcoind_rpc` crate, instead of emitting `(u32, Block)`, we want to emit `(CheckPoint, Block)`. This is because for the first emission (given a start height that is >1), we can no longer rely on the block header to construct the chain update (since we now need to connect to the genesis block!). Additionally, if there is no previous checkpoint (`Emitter::last_cp`), the first emission's checkpoint should also include the genesis block.

**Use case**

This fixes #1107 without requiring methods of `TxGraph` and `IndexedTxGraph` to take in an `Option` of `BlockId` as the *static block*.

The additional benefit of this is that if the caller accidentally updates `LocalChain` with dat from the wrong network, the update can never connect (since the genesis block cannot be replaced).

**Additional context**

Suggested by @evanlinjin in bitcoindevkit/.github#72.

Comment on `PR#1116`: https://github.com/bitcoindevkit/bdk/pull/1116#issuecomment-1752042855

### LINKED ISSUE #1107: BlockId should not have a default and passing a chain tip should be an `Option`
It does not make sense for a BlockId to have a default. We use the `Default` to semantically mean a null block. e.g.
```
self.indexed_graph
            .graph()
            .filter_chain_unspents(
                &self.chain,
                self.chain.tip().map(|cp| cp.block_id()).unwrap_or_default(),
                self.indexed_graph.index.outpoints().iter().cloned(),
            )
```

I think these should just take an `Option` for the tip since that's the rusty way of communicating null. When it's `None` it should just be assumed that nothing is in the chain.
Maybe the tip should take an argument that is `Option<B> where B: Into<BlockId>` to avoid the `map`.

On a related note `ChainOracle` has `get_chain_tip` despite this not being used anywhere. I'd say we should remove it. Perhaps we will want this later but we can just make a new trait for that like `GetChainTip`.

### PR DESCRIPTION
**Pull request by evanlinjin**
closes #1079
closes #1107

### Description

Many methods of `TxGraph` require a `chain_tip: BlockId` input to use against a `ChainOracle` implementation. This is used to ask the `ChainOracle` implementation whether a certain block exists in the chain identified by the `chain_tip`. This guarantees that the `TxGraph` methods will return a consistent history of transactions.

However, the `ChainOracle` trait's `get_chain_tip` method returns an option of `BlockId`. It becomes unclear what to do when `get_chain_tip` returns `None`.

This PR changes the `ChainOracle::get_chain_tip` method to always return a `BlockId` (no `Option`). `LocalChain` now hardwires the genesis block in order to implement `ChainOracle`.

`bdk::Wallet` and `bdk_file_store::Store` are changed to have separate constructor methods for initializing a fresh instance and recovering a previous instance from persistence.

### Changelog notice

- Changed `ChainOracle::get_chain_tip` method to return a `BlockId` instead of an `Option` of a `BlockId`.
- Refactored `LocalChain` so that the genesis `BlockId` is hardwired. This way, the `ChainOracle::get_chain_tip` implementation can always return a tip.
- Add `is_empty` method to `PersistBackend`. This returns true when there is no data in the persistence.
- Changed `Wallet::new` to initialize a fresh wallet only.
- Added `Wallet::load` to restore an instance of a wallet.
- Replaced `Store::new` with separate methods to create/open the database file.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

### CODE REVIEW

File: crates/bdk/README.md
Comment (LLFourn) [inline crates/bdk/README.md]: I'd say either make new functions to support this or create a `Params` type or a `WalletBuilder` or something like that.  It's hard to figure out what these `None`s mean just by looking at the signature.

File: crates/bdk/src/wallet/mod.rs
Comment (LLFourn) [inline crates/bdk/src/wallet/mod.rs]: I think we should consider a explicit API  which distinguishes between initializing a wallet for the first time and loading it -- so `::new` and `::load`. It should be simpler for developers and could avoid this error case. Furthermore consider the implications of also doing this: https://github.com/bitcoindevkit/bdk/issues/1101#issuecomment-1701929749 and we also persist the `Network` in the first changeset (waving hands a bit here). Then load could load everything from in one go without the possibility of contradiction between runtime data and persisted data.

So `new` would take what it takes now but would persist the genesis block, network and keychains immediately. `load` would get everything from persistence and applications would just inspect it at runtime to find out what kind of wallet the user had. If we really want this kind of function we could have `new_or_load` as well which just tries to load, checks everything is consistent with the paramters (throwing errors like above if not) or if it can't be loaded, making a new one.

File: crates/chain/src/genesis.rs
Comment (notmandatory) [inline crates/chain/src/genesis.rs]: Why can't we use `rust-bitcoin` for this?

https://docs.rs/bitcoin/latest/bitcoin/blockdata/constants/fn.genesis_block.html#

Comment (evanlinjin) [reply crates/chain/src/genesis.rs]: I didn't know this existed. 😮

Comment (evanlinjin) [reply crates/chain/src/genesis.rs]: Removed.

File: crates/bdk/README.md
Comment (evanlinjin) [reply crates/bdk/README.md]: Let's do this in a separate PR, once #1101 is completed also.

Comment (LLFourn) [reply crates/bdk/README.md]: Do what? Please don't add another param to this function in this PR.

Comment (evanlinjin) [reply crates/bdk/README.md]: Sorry I misread your initial comment. I created a separate method for now.

File: crates/bdk/src/wallet/mod.rs
Comment (evanlinjin) [reply crates/bdk/src/wallet/mod.rs]: I think `Wallet::new` probably needs a check as to whether the underlying `PersistenceBackend` is empty. Otherwise, there may be accidental overwrites.

Comment (LLFourn) [reply crates/bdk/src/wallet/mod.rs]: I don't think the `PersistenceBackend` has an overwrite API though?

Comment (evanlinjin) [reply crates/bdk/src/wallet/mod.rs]: When `bdk_file_store::Store` is initialized, the logical cursor is positioned after the "magic bytes".

`Wallet::new` does not load any data from the `PersistenceBackend`.

So the next write will start at the position after "magic bytes" (if I'm not mistaken).

`Wallet::new` always writes the initial changeset.

Comment (LLFourn) [reply crates/bdk/src/wallet/mod.rs]: So this sounds like an implementation specific problem. I don't think that `Wallet::new` needs to check anything here. We can just have a safe method in `bdk_file_store` that errors out when you create a new db for a file that already exists. Remember the data we are storing is fine to delete. It can always be reconstructed.

**Comment by evanlinjin** [comment]
I'm really close. Remaining tasks:

- [x] Make `bdk_file_store`'s constructor methods "safe" so we don't overwrite existing valid data. As mentioned here: https://github.com/bitcoindevkit/bdk/pull/1178#discussion_r1374036437
- [x] Documentation and tests for `Wallet`'s constructors.

File: crates/bdk/src/wallet/mod.rs
Comment (evanlinjin) [reply crates/bdk/src/wallet/mod.rs]: I now have separate methods on `bdk_file_store::Store` for initializing and loading.

Comment (LLFourn) [reply crates/bdk/src/wallet/mod.rs]: So I didn't get feedback on the idea of doing the changes I suggested in #1101 as part of this. Is this PR intended to be a stepping stone towards doing #1101 or do you not think #1101 is viable. I doesn't seem valuable to do this change and then change the API again when we do #1101.

File: crates/chain/src/persist.rs
Comment (LLFourn) [inline crates/chain/src/persist.rs]: why does this need to be on the trait. I don't see why we need it.

File: crates/bdk/src/wallet/mod.rs
Comment (evanlinjin) [reply crates/bdk/src/wallet/mod.rs]: ~How about we first move the descriptors to a field in the wallet's ChangeSet. Then in a separate PR, we change the keychain txout index to have changesets with descriptors? This way the API of wallet stays consistent.~

^ The above is a bad idea.

I think it's best to wait for @danielabrozzoni's PR for #1101 and have that merged first and base this work on that.

File: crates/chain/src/persist.rs
Comment (evanlinjin) [reply crates/chain/src/persist.rs]: This is for the `Wallet::new_or_load` method. So that we know whether we need to initialize (where we commit the first changeset) or load from existing. Technically, we can check whether required fields exists (i.e. network type and descriptors)... Would you prefer this way instead?

**Edit:** After thinking about it, I think a better option would be for `PersistBackend::load_from_persistence` method to return `Result<Option<C>, Self::LoadError>`. It is helpful to know the difference between having no data in persistence v.s. only having persisted a `Default` changeset.

File: crates/bdk/src/wallet/mod.rs
Comment (LLFourn) [inline crates/bdk/src/wallet/mod.rs:140]: Should probably `debug_assert` at least that these are not conflicting.

A cleaner less brittle approach would be to have some fixed `Header` data that you can set in persistence. Make an issue if you think it's worth discussing.

**Review by LLFourn — APPROVED** [review]
ACK 632b8067cd173700c04aa1b848da49e13c0dea13

File: crates/bitcoind_rpc/src/lib.rs
Comment (vladimirfomene) [inline crates/bitcoind_rpc/src/lib.rs]: I wonder what we can do so that we don't forget to write this doc.

File: crates/bdk/src/wallet/mod.rs
Comment (vladimirfomene) [inline crates/bdk/src/wallet/mod.rs:405]: What does `empty` here mean? From my understanding, it means the wallet has an empty state (empty IndexedTxGraph, empty localchain). We might want to improve the doc here.

File: crates/bdk/src/wallet/mod.rs
Comment (vladimirfomene) [inline crates/bdk/src/wallet/mod.rs:478]: nits: I will change the name of this method to `init_from_changeset`. Feel free to ignore if it doesn't resonate.

File: crates/file_store/src/store.rs
Comment (vladimirfomene) [inline crates/file_store/src/store.rs:45]: Is it really in write only mode given that the file is also given read option on line 65

**Review by vladimirfomene — COMMENTED** [review]
Great work Evan! I just have some nits.

File: crates/bitcoind_rpc/src/lib.rs
Comment (notmandatory) [reply crates/bitcoind_rpc/src/lib.rs]: Looks like docs here should be something like:
```suggestion
    /// Construct a new [`Emitter`] with the given RPC `client`, `last_cp` and `start_height`.
    ///
    /// `last_cp` is the check point used to find the latest block which is still part of the best chain.
    /// `start_height` is the block height to start emitting blocks from.
```

**Review by notmandatory — APPROVED** [review]
ACK e342a8c3dd5ce7775756df8ae7bbbe2eb3f61863

Looks good, nice to have that chain tip Option removed!

**Comment by notmandatory** [comment]
This should also close #1107 .

**Comment by notmandatory** [comment]
@evanlinjin looks like this one has a conflict so needs a rebase before I can merge it.

**Comment by evanlinjin** [comment]
@notmandatory don't merge yet, I want to address comments and make clippy happier.

**Pull request merged**
